[#14] Update barcode helper

itsmattking opened the pull request:
Breaks out `BarCode` into interface and creates `BarCode39` and `BarCode128`. Further bar code definitions should use `BarCode` interface.

itsmattking committed:
[#14] Update barcode helper

Breaks out `BarCode` into abstract class and creates `BarCode39` and `BarCode128`.

Further bar code definitions should use `BarCode` as the base class.

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode.kt:
```diff
@@ -0,0 +1,9 @@
+import com.sainsburys.k2zpl.command.ZplCommand
+import com.sainsburys.k2zpl.command.options.ZplFieldOrientation
+
+internal abstract class BarCode : ZplCommand {
```
I think we can get away with making this an interface, especially as it's internal. We can always make it an abstract class again later if needed. WDYT?

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode39.kt:
```diff
@@ -1,30 +1,34 @@
-    val line: Int,
-    val lineAbove: ZplYesNo
-) : ZplCommand {
+    val line: ZplYesNo,
```
I see that changing `line` from a constrained `Int` to a `ZplYesNo` if firstly a breaking change and secondly causes the generated ZPL code to say `Y` or `N` rather than a 1-digit number, so I wanted to double check if this was intentional.

jbduncan replied:
Having said this, I see that the command is different - `^B3` rather than `^B1` - so does this change how the Zebra printer interprets things?

itsmattking replied:
The `line` was mistakenly added. It turns out for both barcode39 and 128 the line option is actually to show the line, not set the thickness. Thus the signature is updated.

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode39.kt:
```diff
@@ -34,32 +38,29 @@ internal data class BarCode(
-    barcodeType: ZplBarcodeType = ZplBarcodeType.CODE_39,
     orientation: ZplFieldOrientation = ZplFieldOrientation.NORMAL,
     checkDigit: Boolean = false,
+    height: Int,
```
Ditto re. moving `height` up with the other mandatory parameters.

jbduncan commented on src/test/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode128Test.kt:
```diff
@@ -0,0 +1,111 @@
+            }
+        }
+        it("requires valid parameters") {
+            table(
+                headers("height"),
+                row(32001),
+                row(0),
+            ).forAll { height ->
+                shouldThrow<IllegalArgumentException> {
+                    subject.copy(
+                        height = height
+                    )
+                }
+            }
```
Nice use of tables, well done! 👍

jbduncan commented on src/test/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode128Test.kt:
```diff
@@ -0,0 +1,111 @@
+            result shouldBe "^BCN,10,N,N,N,N"
+        }
+        it("uses orientation parameter properly") {
+            ZplFieldOrientation.entries.forEach {
+                subject.copy(orientation = it).testBuildString() shouldBe "^BC${it.code},10,N,N,N,N"
+            }
```
Have we considered making this a table-driven test, and similarly for all other tests that use `.entries.forEach`?

jbduncan commented on src/test/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode39Test.kt:
```diff
@@ -14,87 +13,87 @@ import io.kotest.data.row
         }
         it("uses orientation parameter properly") {
             ZplFieldOrientation.entries.forEach {
-                barCode.copy(orientation = it).testBuildString() shouldBe "^B1${it.code},N,10,7,Y"
+                subject.copy(orientation = it).testBuildString() shouldBe "^B3${it.code},N,10,N,N"
             }
```
Ditto re. table-driven tests.

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode39.kt:
```diff
@@ -34,32 +38,29 @@ internal data class BarCode(
- * @param barcodeType Barcode type
  * @param orientation The orientation of the barcode.
- * @param checkDigit Whether to include a check digit.
+ * @param checkDigit Mod-43 check digit
```
The previous `@param` description "Whether to include a check digit." was pretty clear to me. I admit I have no context on the ZPL programming language, but have we considered merging the old and new descriptions together like so?

```kotlin
/**
 * @param checkDigit whether to include a Mod-43 check digit
 */
```

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode128.kt:
```diff
@@ -0,0 +1,75 @@
+    )
+}
+
+/**
+ * Creates a Code 128 barcode marker
+ * @param data data encoded in the barcode
+ * @param x horizontal position
+ * @param y vertical position
+ * @param orientation The orientation of the barcode.
+ * @param height The height of the barcode.
+ * @param interpretationLine print interpretation line
+ * @param lineAbove print interpretation line above code
+ * @param checkDigit UCC check digit
+ * @param mode barcode mode
+ */
```
Minor nit: I see that some `@param`s end with full stops whereas the others don't, and that some start with a capital letter and others don't.

I believe common Javadoc (and therefore KDoc?) practice is to not start with capitals, have no full stops and separate "sentences" with semi-colons.

But I'm happy to go for any style as long as we're consistent!

What do you think?

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode128.kt:
```diff
@@ -0,0 +1,75 @@
+ * @param height The height of the barcode.
+ * @param interpretationLine print interpretation line
+ * @param lineAbove print interpretation line above code
+ * @param checkDigit UCC check digit
```
Let's consider fleshing this `@param` description similarly to my other comment for `barcode39` with something like:

```
/**
 * @param checkDigit whether to include a UCC check digit
 */
```

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode39.kt:
```diff
@@ -34,32 +38,29 @@ internal data class BarCode(
- * @param lineAbove Whether to include a line above the barcode.
+ * @param interpretationLine print interpretation line
+ * @param lineAbove print interpretation line above code
  */
```
Similarly to my other comment, I believe common Javadoc (and therefore KDoc) practice for `@param` tags is to not start with capitals, have no full stops and separate "sentences" with semi-colons.

jbduncan commented on src/main/kotlin/com/sainsburys/k2zpl/command/barcode/BarCode128.kt:
```diff
@@ -0,0 +1,75 @@
+    x: Int,
+    y: Int,
+    orientation: ZplFieldOrientation = ZplFieldOrientation.NORMAL,
+    height: Int,
```
I'm tempted to move `height` up with the other mandatory parameters so they're all one place.

If we didn't do this, I'm not sure how the Kotlin compiler would interpret things if we accidentally did the following:

```kotlin
BarCode39(
    "1234567890",
    10,
    10,
    10 // Does it know that this is the height?
)
```

What do you think?

jbduncan reviewed (commented):
I have a few stylistic suggestions, but otherwise, API-wise this looks good to me! 👍

itsmattking committed:
[#14] Update barcode helper

PR feedback:

- Make `BarCode` an `interface`
- Use `table` entries for tests
- Doc consistency fixes
